python/taichi/lang/shell.py

```diff
@@ -0,0 +1,221 @@
+import sys, os, atexit
+
+class ShellType:
+    NATIVE = 'Python shell'
+    IDLE = 'Python IDLE shell'
+    IPYTHON = 'IPython TerminalInteractiveShell'
+    JUPYTER = 'IPython ZMQInteractiveShell'
+    IPYBASED = 'IPython Based Unknown Shell'
+    SCRIPT = None
+
+def _show_idle_error_message():
+    import taichi as ti
+    ver = sys.version[:3]
+    if ti.get_os_name() == 'win':
+        # For Windows Python IDLE: sys.executable = "../pythonw.exe"
+        path = os.path.join(os.path.abspath(sys.executable), f'lib/python{ver}/code.py')
+    else:
+        path = f'/lib/python{ver}/code.py'
+    print('It\'s detected that you are using Python IDLE in **interactive mode**.')
+    print('However, Taichi could not be fully functional due to IDLE limitation, sorry :(')
+    print('Either run Taichi directly from script, or use IPython notebook instead.')
+    print('We do care about your experience, no matter which shell you prefer to use.')
+    print('So, if you would like to play with Taichi in your favorite IDLE, we may do a dirty hack:')
+    print(f'Open "{path}" and add the following line to `InteractiveInterpreter.runsource`, right below `# Case 3`:')
+    print('''
+class InteractiveInterpreter:
+    ...
+
+    def runsource(self, source, filename="<input>", symbol="single"):
+        ...
+
+        # Case 3
+        (lambda o,k:o.path.exists(k+'ppid_'+str(o.getpid()))and(lambda f:(f.write(f'\\n===\\n'+source),f.close()))(open(k+'source','a')))(__import__('os'),'.tmp_idle_') # Add this line!
+        self.runcode(code)
+        return False
+
+    ...
+
+        ''')
+    print('Then, restart IDLE and enjoy, the sky is blue and we are wizards!')
+
+
+def get_shell_name():
+    shell = os.environ.get('TI_SHELL_TYPE')
+    if shell is not None:
+        return getattr(ShellType, shell.upper())
```

Comment:
If they encounter any problem detecting shell type, they can still use a env var to force which to wrap.